### ISSUE TITLE
[DependencyInjection] mutated service loses its constructor arguments

### ISSUE THREAD

**Issue by MatTheCat** [symfony/symfony#23325]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3.2

We have a dependency to a bundle which define this service:

```xml
<service
    id="dnl_user.user_registrar"
    class="Dnl\Bundle\UserBundle\Security\User\Registrar"
>
    <argument type="service" id="security.password_encoder"/>
    <argument type="service" id="dnl_user.user_entity_manager"/>
    <argument type="service" id="event_dispatcher"/>
</service>
```

We override this definition through a compiler pass:

```php
$container
    ->getDefinition('dnl_user.user_registrar')
    ->setClass('Dnl\BackBundle\User\Registrar')
;
```

Now when injecting the service into a controller action we get

- when typehinting with `Dnl\BackBundle\User\Registrar`

> Argument 1 passed to Dnl\Bundle\UserBundle\Security\User\Registrar::__construct() must implement interface Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface, none given

- when typehinting with `Dnl\Bundle\UserBundle\Security\User\Registrar`

> Controller "Dnl\BackBundle\Controller\UserAdminController::newAction()" requires that you provide a value for the "$registrar" argument.

Setter injection seems to be working except for overridden services, is there something I overlooked?

**Comment by stof** [comment]
The code you are showing is not overwriting the service definition. It is mutating the existing definition.

Overwriting the service indeed looses all constructor arguments (and also all method calls), as it replaces the whole Definition object. But this does not look like what you're doing here.

to help you with your errors, I would need the stack traces too (or at least the location of the error), to know where they happen.

**Comment by MatTheCat** [comment]
Here you go:

![mlechat base project dev7 ovh-user-admin-new iphone 5](https://user-images.githubusercontent.com/1898254/27691907-af33a5b0-5ce5-11e7-8978-9eb4db44e98e.png)

**Comment by stof** [comment]
OK, so the first issue is because Symfony's autowiring looked for a `Dnl\BackBundle\User\Registrar` service id (as this is the class it tried to autowire). As it could not find one and the class is in the same top-level namespace than the autowired class (your controller), it created a new service. and this is the one missing the constructor argument. The error does **not** come from your `dnl_user.user_registrar` service.
I don't understand why the newly created service does not get autowired properly though. @nicolas-grekas @weaverryan do you have any idea ?


If you want your `dnl_user.user_registrar` service to be used when autowiring the `Dnl\BackBundle\User\Registrar` typehint, create a private alias named `Dnl\BackBundle\User\Registrar` and targetting `dnl_user.user_registrar`.

**Comment by MatTheCat** [comment]
So I'm not sure, is the issue from my side because I didn't define an alias or from Symfony's side because this should work?

**Comment by stof** [comment]
Well, both.
If you want your existing service to be used, you need the alias (and in your case, I'm almost sure you want your own service to be used).
But when Symfony creates one, it must configure it properly (or fail loudly if it cannot). Defining a broken service is not good. So I will keep this issue opened at the moment as we need to figure out what's wrong with it.

**Comment by MatTheCat** [comment]
Ok I understand thanks for the explanation!

**Comment by nicolas-grekas** [comment]
The current WTF will be fixed by #23350. I guess we can close then?

**Comment by javiereguiluz** [comment]
Yes, it looks like we can close this issue. Thanks!
